keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,305 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Scalability and performance of kube-apiserver is crucial for scalability
+of the whole Kubernetes cluster. Given that kube-apiserver is cpu-intensive
+process, scaling a single instance of it translates to optimizing amount
+of work is needed to process a request (cpu cycles and amount of allocated
+memory, as memory management is significant part of work done be
+kube-apiserver).
+
+This proposal is aiming to significantly reduce amount of work spent on
+serializing objects as well as amount of allocated memory to process that.
+
+## Motivation
+
+Running different types of scalability tests and analyzing large production
+clusters proves that large number of watches watching the same set of objects
+may cause significant load on kube-apiserver. An extreme example of it is
+[#75294][], where creation of a single large Endpoints object (almost 1MB of
+size, due to 5k pods backing it) in 5k-node cluster can completely overload
+kube-apiserver for 5 seconds.
+
+The main reason for that is that for every watcher (Endpoints are being watched
+by kube-proxy running on every one) kube-apiserver independently serializes
+(which also requires deep-copy) every single object being send via this watch.
+
+While this problem is extremely visible for watch, the situation looks the same
+for regular GET/LIST operations - reading the same object N times will result
+in serializing that N times independently.
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver and number of memory allocations, by avoiding
```

Comment:
Protobuf encoding should only result in 1 or 2 memory allocations.  Removing that is unlikely to show any significant benefit.

Note that the protobuf encoding changes proposed by @apelisse should have a significant speed up in protobuf encoding, which means that we should really consider the impact of this change AFTER we get those changes in mainline.  It may improve things - or it may have much less impact now that protobuf encoding is significantly faster.

Reply:
There is still a deep-copy issue that we're doing for every single watch event. So that helps, but there is still a lot of place for improvement.